### DISCUSSION TITLE
How to send file using FTP

### DISCUSSION THREAD

**Discussion by Ricmelis80** [micropython/micropython#12649] [Core Development]
Hy everyone!

I have  connected my Raspberry Pico W over Wifi with an AP connected to a company server. 

Since I have been enabled to use only the FPT protocol, how I can transfer my sensor data storage ( in txt format) to a shared folder of such server?   

Thanks in advance!

**Comment by robert-hh** [comment]
There is a ftp library which I tested some time ago and found it working well. https://github.com/SpotlightKid/micropython-ftplib

**Reply by Ricmelis80** [reply]
Hy Robert,

I have just tested this library in my Pico W  but I have encountered the following problem ( also when I launch the "ftpsample.py" as is)  : 

"Traceback (most recent call last):
File "", line 4, in
File "ftplib.py", line 160, in init
File "ftplib.py", line 217, in connect
File "ftplib.py", line 180, in _create_connection
File "ftplib.py", line 71, in _resolve_addr
OSError: -6"

The author of library said that  Pico W is not tested or supported by this codes... You have tested it with Pico W model?

**Reply by robert-hh** [reply]
I used that one years ago with an ESP8266, not RPI pico. So it seems it has to be adapted for it.

**Reply by Ricmelis80** [reply]
Yes I suppose it

**Reply by scruss** [reply]
The author of the library tells me that the Pico W is now tested and supported: [fix: overhaul FTP over TLS API · SpotlightKid/micropython-ftplib@6c19a84](https://github.com/SpotlightKid/micropython-ftplib/commit/6c19a8446c0c0423267ecb79a949e14f05efb613)

**Comment by steveh127** [comment]
I have been trying the ftplib library with a Pico W and it's working 'out the box' for me but only to connect to an anonymous FTP server. Attempting to connect using a named account only failed with a 'login details incorrect' type error so I was talking to the FTP server.

The only imported module used by ftplib library is usocket so there really should be no issue connecting with any network capable MCU. 

A more radical solution would be to set up your Pico W as an FTP server. See [https://github.com/cpopp/MicroFTPServer](url). Again this worked on a Pico W without any modifications. Just connect to a network and import uftp. It's again only an anonymous FTP server but I can connect to it using FileZilla and from another MCU using ftplib.

**Reply by robert-hh** [reply]
A more recent and improved version of that ftp server is here: https://github.com/robert-hh/FTP-Server-for-ESP8266-ESP32-and-PYBD
It supports a background mode.

**Comment by nikitamane1999** [comment]


**Reply by robert-hh** [reply]
If may be this line 353 in ftplib.py:
`                line = line + (b'\n' + nextline)`

Try changing it to:
`                line = line + ('\n' + nextline)`

**Comment by nikitamane1999** [comment]


**Reply by robert-hh** [reply]
> Error: Could not connect to ('172.16.1.55', 21)

That's a different error. Can you connect with an ftp client from the PC?. Windows has an ftp commandline version as well.
